go/mcap/lexer.go

```diff
@@ -191,16 +188,43 @@ func (l *Lexer) Next(p []byte) (TokenType, []byte, error) {
 		if l.maxRecordSize > 0 && recordLen > uint64(l.maxRecordSize) {
 			return TokenError, nil, ErrRecordTooLarge
 		}
-		if opcode == OpChunk && !l.emitChunks {
-			err := loadChunk(l, recordLen)
-			if err != nil {
-				if l.emitInvalidChunks {
-					var invalidCrc *errInvalidChunkCrc
-					if errors.As(err, &invalidCrc) {
-						return TokenInvalidChunk, nil, err
+
+		// Chunks and attachments require special handling to avoid
+		// materialization into RAM. If it's a chunk, open up a decompressor and
+		// swap it in as the active reader, then continue on the next message
+		// (which will be from the chunk data). If it's an attachment, parse the
+		// record into an AttachmentReader and call any user-supplied callback.
+		// Then discard any remaining data and continue to the next record.
+		switch opcode {
+		case OpChunk:
+			if !l.emitChunks {
+				err := loadChunk(l, recordLen)
+				if err != nil {
+					if l.emitInvalidChunks {
+						var invalidCrc *errInvalidChunkCrc
+						if errors.As(err, &invalidCrc) {
+							return TokenInvalidChunk, nil, err
+						}
 					}
+					return TokenError, nil, err
+				}
+				continue
+			}
+		case OpAttachment:
+			limitReader := io.LimitReader(l.reader, int64(recordLen))
+			attachmentReader, err := parseAttachmentReader(limitReader)
+			if err != nil {
+				return TokenError, nil, fmt.Errorf("failed to parse attachment: %w", err)
+			}
+			if l.attachmentCallback != nil {
+				err := l.attachmentCallback(attachmentReader)
+				if err != nil {
+					return TokenError, nil, fmt.Errorf("failed to handle attachment: %w", err)
 				}
-				return TokenError, nil, err
+			}
+			_, err = io.Copy(io.Discard, limitReader)
```

Comment:
can we check if this is a Seeker and seek ahead instead if possible?

Reply:
yep, will do

Reply:
added that